Remove atomic enum election::state_m and instead require locking election::mutex.

clemahieu opened the pull request.

pwojcikdev committed:
Fix AEC insert race condition

clemahieu committed:
Move evaluation of election->confirmed () within debug_assert so it doesn't need to be evaluated when debugging is off.

clemahieu committed:
Remove atomic enum election::state_m and instead require locking election::mutex.

clemahieu committed:
Removing erasing recently_confirmed

dsiganos commented:
@clemahieu, could you explain why this change is needed, what problem does it solve.

clemahieu commented:
This change is removing the atomic election::state_m variable in favor of using the election mutex for synchronization. The issue is there are two synchronization mechanisms in use, locking the mutex or doing an atomic compare exchange, which has race conditions.

From what I recall the election can be doing an operation with a locked mutex, and a separate action can change the election state because it doesn't respect the mutex being locked.

dsiganos committed:
Fix unit test rep_crawler.recently_confirmed

dsiganos committed:
Remove active_transactions::insert_impl and remove the passing of mutex

A mutex was needlessly being passed from function to function and a
function wrapper existed needlessly.

dsiganos committed:
Return early, if stopped, to remove a large level of indentation.

dsiganos committed:
Break up long complex statement into 2 statements

dsiganos committed:
Refactor function to remove unnecessary lock/unlocks and make it readable

The critical section is now obvious and easy to see

clemahieu commented:
LGTM
